File: src/test/resources/org/bau/benchmarks/munchausen.py
# https://github.com/jabbalaci/SpeedTests
N = 440_000_000

def main():
    cache: list[int] = get_cache()
    for n in range(0, N):
        if is_munchausen(n, cache):
            print(n)

def is_munchausen(number: int, cache: list[int]) -> bool:
    n = number
    total = 0
    while n > 0:
        digit = n % 10
        total += cache[digit]
        if total > number:
            return False
        n = n // 10
    return total == number

def get_cache() -> list[int]:
    return [0] + [i**i for i in range(1, 9 + 1)]

main()

File: src/test/resources/org/bau/benchmarks/munchausen.rs
// https://github.com/jabbalaci/SpeedTests

const N: i32 = 440_000_000;

fn main() {
    let cache = get_cache();
    for n in 0..N {
        if is_munchausen(n, &cache) {
            println!("{}", n);
        }
    }
}

fn is_munchausen(number: i32, cache: &[i32; 10]) -> bool {
    let mut n = number;
    let mut total = 0;
    while n > 0 {
        let digit = n % 10;
        total += cache[digit as usize];
        if total > number {
            return false;
        }
        n /= 10;
    }
    number == total
}

fn get_cache() -> [i32; 10] {
    let mut cache = [0; 10];
    for n in 1..=9 {
        cache[n] = (n as i32).pow(n as u32);
    }
    cache
}
